File: csharp/50.pow-x-n.cs
/*
 * @lc app=leetcode id=50 lang=csharp
 *
 * [50] Pow(x, n)
 */

 System.Console.WriteLine(new Solution().MyPow(2, -2147483648));

// @lc code=start
public class Solution {
    public double MyPow(double x, int n) {
        if (n == 0 || x == 1) {
            return 1;
        }
        else if (n > 0) {
            if (n == 1)
                return x;

            return (n%2 == 0 ? 1 : x) * MyPow(x*x, n/2);
        }
        else {
            if (n == -1)
                return 1/x;

            return (n%2 == 0 ? 1 : 1/x) * MyPow(x*x, n/2);
        }
    } 
}
// @lc code=end

File: cpp/50.pow-x-n.cpp
/*
 * @lc app=leetcode id=50 lang=cpp
 *
 * [50] Pow(x, n)
 */


// @lc code=start
class Solution {
public:
    double myPow(double x, int n) {
        if (n == 0 || x == 1)
            return 1;
            
        if (n < 0)
            return _myPow(1/x, -n);

        return _myPow(x, n);
    }
private:
    double _myPow(double x, int n, double prod = 1) {
        return n == 0 ? prod : _myPow(x, n-1, x*prod);
    }
};
// @lc code=end
